payload-receiver-service/tests/integration/smoke_tests.py

```diff
@@ -0,0 +1,288 @@
+from nats.aio.errors import ErrConnectionClosed, ErrNoServers, ErrTimeout
+from opni_nats import NatsWrapper
+
+kube_fname = "kube.yaml"
+nw = NatsWrapper()
+fake = Faker()
+port_forward_command = "kubectl --kubeconfig kube.yaml -n opni-demo port-forward "
+json_payload = []
+
+def test_prs_happy_path():
+    
+    # This test is to verify the happy path functionality of the Payload Receiver Service (PRS). 
+    # In this test, we are verifying that each of the following fields are successfully added to a log submitted to the PRS.
+        # time
+        # window_dt
+        # window_start_time_ns
+        # _id
+
+    tr_queue = Queue()
+
+    start_process("killall kubectl")
+    
+    log_data = ('{"log": {"0":"' + fake.sentence(10) + '"}}')
+
+    kube_nats_process = Popen([port_forward_command + "svc/nats-client 4222:4222"], shell=True)
+    kube_payload_process = Popen([port_forward_command + "svc/payload-receiver-service 8080:80"], stdout=PIPE, shell=True)
+        
+    i = 0
+    while i <= 5: 
+        if kube_payload_process.poll != None and kube_nats_process.poll != None :
+            time.sleep(1)
+            i += 1
+        else :
+            raise Exception('kube_payload_process and/or kube_nats_process are not running')
+
+    # nats subscribe
+    t = subscribe(tr_queue, log_data)
+    
+    wait_for_seconds(2)
+
+    print('Sending Dataset')
+    r = requests.post('http://localhost:8080',
+            data=log_data,
+            verify=False)
+        
+    if len(r.content) != 0:
+        print(('Request Content: '), r.content)
+        print(('Request Headers: '), r.headers)
+        print(('Request Status Code'), r.status_code)
+        bad_content = '{"detail":"Something wrong with request'
+        content = r.content.decode("utf-8")
+        if bad_content in content:
+            raise Exception('Bad Request sent to API')
+        if r.status_code != 200:
+            raise Exception('Bad Request sent to API')
+    
+    wait_for_seconds(2)
+
+    loop = asyncio.get_event_loop()
+    loop.stop()
+
+    kube_payload_process.terminate
+    kube_payload_process.wait
+    kube_nats_process.terminate
+    kube_nats_process.wait
+    assert kube_payload_process.returncode == None
+    assert kube_nats_process.returncode == None
+
+    global json_payload
+    assert json_payload["time"]["0"] != None
+    assert json_payload["window_dt"]["0"] > 1632792570000
+    assert json_payload["window_start_time_ns"]["0"] > 1632792570000000000
+    assert (int(json_payload["_id"]["0"])) > 16327923057044440000000000000000000
+
+def test_prs_unique_id():
+    
+    # This test is to verify the each log submitted to the Payload Receiver Service (PRS) is unique. 
+
+    tr_queue = Queue()
+
+    start_process("killall kubectl")
+    
+    global json_payload
+    log_data = ('{"log": {"0":"' + fake.sentence(10) + '"}}')
+
+    kube_nats_process = Popen([port_forward_command + "svc/nats-client 4222:4222"], shell=True)
```

Comment:
These setup steps are duplicated throughout tests, look into Pytest fixtures to simplify this.

Reply:
I resolved this with fixtures.